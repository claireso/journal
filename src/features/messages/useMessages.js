import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

const MessagesContext = React.createContext()

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

export const MessagesProvider = ({ children }) => {
  const [state, setState] = useState([])

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = () => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status) => (message) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
