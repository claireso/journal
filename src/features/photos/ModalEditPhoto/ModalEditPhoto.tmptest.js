//@TODO restore test

// import { render, fireEvent, act } from '@testing-library/react'

// import Edit from './Edit'

// import PhotosReducer from '../../reducer'

// const { PhotosProvider, INITIAL_STATE } = PhotosReducer

// const { usePhotosReducer: originalUsePhotosReducer } = PhotosReducer

// describe('<Edit />', () => {
//   const renderComponent = () => {
//     return render(
//       <PhotosProvider value={INITIAL_STATE}>
//         <Edit id={199} />
//       </PhotosProvider>
//     )
//   }

//   const modelPhoto = (photo) => ({ ...photo, source: `/uploads/${photo.name}` })
//   const single = modelPhoto(global.__PHOTOS__.items[0])

//   afterEach(() => {
//     PhotosReducer.usePhotosReducer = originalUsePhotosReducer
//   })

//   test('should load resource', () => {
//     const spyLoad = jest.fn()

//     PhotosReducer.usePhotosReducer = () => [INITIAL_STATE, { loadResource: spyLoad }]

//     renderComponent()

//     expect(spyLoad).toHaveBeenCalledWith(199)
//   })

//   test('should load resource (mismatch photo id)', () => {
//     const spyLoad = jest.fn()

//     PhotosReducer.usePhotosReducer = () => [
//       { ...INITIAL_STATE, single: modelPhoto(global.__PHOTO__) },
//       { loadResource: spyLoad }
//     ]

//     renderComponent()

//     expect(spyLoad).toHaveBeenCalledWith(199)
//   })

//   test('should fill form with photo', async () => {
//     const spyLoad = jest.fn()

//     PhotosReducer.usePhotosReducer = () => [
//       {
//         ...INITIAL_STATE,
//         single: single
//       },
//       { loadResource: spyLoad }
//     ]

//     const { container } = renderComponent()

//     expect(spyLoad).not.toHaveBeenCalled()

//     //@TODO: test background color
//     expect(container.querySelector('input[name="title"]').value).toEqual('')
//     expect(container.querySelector('input[name="description"]').value).toEqual('Février 2019')
//     expect(container.querySelector('img').src).toEqual('http://localhost/uploads/01d2y7jt2j24dv0s82m9xq729d.jpg')
//     expect(container.querySelector('select').value).toEqual('right')
//   })

//   test('should submit form', async () => {
//     const spySubmit = jest.fn()
//     PhotosReducer.usePhotosReducer = () => [
//       {
//         ...INITIAL_STATE,
//         single: single
//       },
//       { editResource: spySubmit }
//     ]

//     const { container, getByLabelText } = renderComponent()

//     fireEvent.change(getByLabelText(/title/i), {
//       target: { value: 'Photo title' }
//     })
//     fireEvent.change(getByLabelText(/description/i), {
//       target: { value: 'Photo description' }
//     })

//     const inputFile = container.querySelector('input[type="file"]')

//     await act(async () => {
//       await fireEvent.change(inputFile, {
//         target: {
//           files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
//         }
//       })
//     })

//     fireEvent.change(getByLabelText(/position/i), { target: { value: 'left' } })

//     fireEvent.click(container.querySelector('input[type="submit"]'))

//     const expectedResult = new FormData(container.querySelector('form'))

//     expect(spySubmit).toHaveBeenCalledWith(199, expectedResult)
//   })

//   test('should not submit form', async () => {
//     PhotosReducer.usePhotosReducer = () => [{ ...INITIAL_STATE, status: 'pending', single: single }, {}]

//     const { container } = renderComponent()

//     expect(container.querySelector('input[type="submit"]')).toBeNull()
//   })
// })
