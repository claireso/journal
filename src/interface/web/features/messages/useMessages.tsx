'use client'

import React, { useState, useContext } from 'react'
import PropTypes from 'prop-types'

interface Message {
  key?: string
  status: 'default' | 'success' | 'error' | 'info'
  message: string
}

interface Actions {
  displaySuccessMessage: (data: Omit<Message, 'status'>) => void
  displayErrorMessage: (data: Omit<Message, 'status'>) => void
  closeMessage: (index: number) => void
  closeAllMessages: () => void
}

const INITIAL_STATE: Message[] = []

type TMessagesContext<T> = [T, React.Dispatch<React.SetStateAction<T>>]

const MessagesContext = React.createContext<TMessagesContext<Message[]>>([null!, null!])

interface MessagesProviderProps {
  children: React.ReactNode
}

export const MessagesProvider = ({ children }: MessagesProviderProps) => {
  const [state, setState] = useState(INITIAL_STATE)

  return <MessagesContext.Provider value={[state, setState]}>{children}</MessagesContext.Provider>
}

const useMessagesContext = () => {
  const context = useContext(MessagesContext)

  if (context === undefined) {
    throw new Error('useMessagesContext must be used within a MessagesProvider')
  }

  return context
}

MessagesProvider.propTypes = {
  children: PropTypes.any
}

const useMessages = (): [Message[], Actions] => {
  const [messages, setMessages] = useMessagesContext()

  const closeMessage = (index: number) => {
    setMessages([...messages.slice(0, index), ...messages.slice(index + 1)])
  }

  const closeAllMessages = () => setMessages([])

  const displayMessage = (status: Message['status']) => (message: Omit<Message, 'status'>) => {
    let index = messages.length

    if (message.key) {
      const newIndex = messages.findIndex((_message: Message) => _message.key === message.key)

      if (newIndex > -1) {
        index = newIndex
      }
    }

    setMessages([
      ...messages.slice(0, index),
      {
        status,
        ...message
      },
      ...messages.slice(index + 1)
    ])
  }

  const displaySuccessMessage = displayMessage('success')
  const displayErrorMessage = displayMessage('error')

  return [
    messages,
    {
      displaySuccessMessage,
      displayErrorMessage,
      closeMessage,
      closeAllMessages
    }
  ]
}

export default useMessages
