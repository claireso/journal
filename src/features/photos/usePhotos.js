import { useQuery, useMutation, useQueryClient } from 'react-query'

import * as api from '@services/api'
import useMessages from '@features/messages/useMessages'

const formatPhoto = (photo) => ({
  ...photo,
  source: `/uploads/${photo.name}`
})

const initialState = {
  items: [],
  pager: {}
}

const CACHE_KEY_LIST = 'photos'
const CACHE_KEY_DETAIL = 'photo'

/**
 * Edit photo
 * @param {object} filters
 * @returns
 */
export const useEditPhoto = (filters = {}) => {
  const queryClient = useQueryClient()
  const [, { displaySuccessMessage, displayErrorMessage }] = useMessages()

  return useMutation(({ id, data }) => api.editPhoto(id, data), {
    onSuccess(photo, { id }) {
      queryClient.setQueryData([CACHE_KEY_DETAIL, id], formatPhoto(photo))

      queryClient.setQueryData([CACHE_KEY_LIST, filters], (photos) => ({
        ...photos,
        items: photos.items.map((_photo) => {
          if (_photo.id !== id) return _photo
          return formatPhoto(photo)
        })
      }))

      displaySuccessMessage({
        key: 'CRUD_PHOTO',
        message: 'Your photo has been updated successfully'
      })
    },
    onError() {
      displayErrorMessage({
        key: 'CRUD_PHOTO',
        message: 'An error has occured during the deletion. Please retry'
      })
    }
  })
}

/**
 * Delete photo and update cache
 * @returns object
 */
export const useDeletePhoto = (filters) => {
  const queryClient = useQueryClient()
  const [, { displaySuccessMessage, displayErrorMessage }] = useMessages()

  return useMutation((id) => api.deletePhoto(id), {
    onSuccess(data, id) {
      queryClient.setQueryData([CACHE_KEY_LIST, filters], (photos) => ({
        items: photos.items.filter((photo) => photo.id !== id),
        pager: {
          ...photos.pager,
          count: photos.pager.count - 1
        }
      }))
      displaySuccessMessage({
        key: 'CRUD_PHOTO',
        message: 'Your photo has been deleted successfully'
      })
    },
    onError() {
      displayErrorMessage({
        key: 'CRUD_PHOTO',
        message: 'An error has occured during the deletion. Please retry'
      })
    }
  })
}

/**
 * Create photo and update cache
 * @returns object
 */
export const useCreatePhoto = (filters = {}) => {
  const queryClient = useQueryClient()
  const [, { displaySuccessMessage, displayErrorMessage }] = useMessages()

  return useMutation((data) => api.createPhoto(data), {
    onSuccess(photo) {
      if (filters.page === '1') {
        queryClient.setQueryData([CACHE_KEY_LIST, filters], (photos) => ({
          items: [formatPhoto(photo), ...photos.items],
          pager: {
            ...photos.pager,
            count: photos.pager.count + 1
          }
        }))
      }
      displaySuccessMessage({
        key: 'CRUD_PHOTO',
        message: 'Your photo has been created successfully'
      })
    },
    onError() {
      displayErrorMessage({
        key: 'CRUD_PHOTO',
        message: 'An error has occured during the creation. Please retry'
      })
    }
  })
}

/**
 * Fetch photos
 * @param {object} queries
 * @param {object} options
 * @returns object
 */
export const usePhotos = (filters = { page: '1' }, options = {}) => {
  const getPhotos = async ({ signal }) => {
    const response = await api.getPhotos(filters.page, { signal })
    return {
      ...response,
      items: response.items.map(formatPhoto)
    }
  }

  const queryOptions = {
    placeholderData: initialState,
    useErrorBoundary: true,
    ...options
  }

  return useQuery([CACHE_KEY_LIST, filters], getPhotos, queryOptions)
}

/**
 * Fetch photo
 * @param {number} id photoId
 * @param {object} options query option
 * @returns
 */
export const usePhoto = (id, options = {}) => {
  const queryClient = useQueryClient()

  const getPhoto = async ({ signal }) => {
    const response = await api.getPhoto(id, { signal })
    return formatPhoto(response)
  }

  const queryOptions = {
    initialData: () => {
      const queries = queryClient.getQueriesData({ active: true })
      const photoQuery = queries.find(([filters]) => filters[0] === CACHE_KEY_LIST)
      return photoQuery?.[1]?.items.find((photo) => photo.id === id)
    },
    initialDataUpdatedAt: () => queryClient.getQueryState({ active: true })?.dataUpdatedAt,
    useErrorBoundary: true,
    ...options
  }

  return useQuery([CACHE_KEY_DETAIL, id], getPhoto, queryOptions)
}
