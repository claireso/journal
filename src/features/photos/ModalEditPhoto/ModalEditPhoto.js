import { useCallback, memo } from 'react'
import PropTypes from 'prop-types'

import { Heading1 } from '@components/Headings'
import FormPhoto from '../FormPhoto'

import { usePhoto } from '@features/photos/usePhotos'

const ModalEditPhoto = ({ id, onSubmit, isProcessing }) => {
  const { data: photo } = usePhoto(id)

  const handleSubmit = useCallback(
    (data) => {
      onSubmit({ id: id, data })
    },
    [id, onSubmit]
  )

  if (!photo) return null

  return (
    <>
      <Heading1>Edit photo</Heading1>
      <FormPhoto onSubmit={handleSubmit} photo={photo} isProcessing={isProcessing} />
    </>
  )
}

ModalEditPhoto.propTypes = {
  id: PropTypes.number.isRequired,
  onSubmit: PropTypes.func,
  isProcessing: PropTypes.bool
}

ModalEditPhoto.defaultProps = {
  onSubmit: () => {},
  isProcessing: false
}

export default memo(ModalEditPhoto)
