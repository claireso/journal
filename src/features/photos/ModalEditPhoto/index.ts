import ModalEditPhoto from './ModalEditPhoto'

export default ModalEditPhoto
