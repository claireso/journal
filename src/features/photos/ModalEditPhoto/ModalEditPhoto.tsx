import { useCallback, memo } from 'react'

import { Heading1 } from '@components/Headings'
import FormPhoto from '../FormPhoto'

import { usePhoto } from '@features/photos/usePhotos'

interface ModalEditPhotoProps {
  id: number
  onSubmit: (data: { id: number; data: FormData }) => void
  isProcessing?: boolean
}

const ModalEditPhoto = ({ id, onSubmit, isProcessing = false }: ModalEditPhotoProps) => {
  const { data: photo } = usePhoto(id)

  const handleSubmit = useCallback(
    (data: FormData) => {
      onSubmit({ id: id, data })
    },
    [id, onSubmit]
  )

  if (!photo) return null

  return (
    <>
      <Heading1>Edit photo</Heading1>
      <FormPhoto onSubmit={handleSubmit} photo={photo} isProcessing={isProcessing} />
    </>
  )
}

export default memo(ModalEditPhoto)
