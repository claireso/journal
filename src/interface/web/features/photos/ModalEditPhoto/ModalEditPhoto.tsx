import { useCallback, memo } from 'react'

import { usePhoto } from '@web/features/photos/usePhotos'
import { Heading1 } from '@web/components/Headings'
import { Loader } from '@web/components/Loader'
import { Photo } from '@models'
import FormPhoto from '../FormPhoto'

interface ModalEditPhotoProps {
  id: Photo['id']
  onSubmit: (data: { id: number; data: Partial<Photo> }) => void
  onCancel: () => void
  isProcessing?: boolean
}

const ModalEditPhoto = ({ id, onSubmit, onCancel, isProcessing = false }: ModalEditPhotoProps) => {
  const { data: photo, error, isFetched } = usePhoto(id)

  const handleSubmit = useCallback(
    (data: Partial<Photo>) => {
      onSubmit({ id: id, data })
    },
    [id, onSubmit]
  )

  if (!isFetched) {
    return <Loader />
  }

  if (error?.response.status === 404) {
    onCancel()
    return null
  }

  return (
    <>
      <Heading1>Edit photo</Heading1>
      <FormPhoto onSubmit={handleSubmit} photo={photo} isProcessing={isProcessing} />
    </>
  )
}

export default memo(ModalEditPhoto)
