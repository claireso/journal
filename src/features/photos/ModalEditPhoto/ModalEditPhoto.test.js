import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'

import ModalEditPhoto from './ModalEditPhoto'

import { PhotosProvider } from '../usePhotos'
import { MessagesProvider } from '../../messages/useMessages'

import * as api from '@services/api'

describe('<ModalEditPhoto />', () => {
  const formatPhoto = (photo) => ({
    ...photo,
    source: `/uploads/${photo.source}`
  })

  const renderComponent = ({ detail, ...props } = {}) => {
    const data = {
      photos: {
        items: global.__PHOTOS__.items.map(formatPhoto),
        pager: global.__PHOTOS__.pager
      },
      isLoading: false,
      isProcessing: false,
      detail: detail || null
    }

    return render(
      <MessagesProvider>
        <PhotosProvider value={data}>
          <ModalEditPhoto id={198} {...props} />
        </PhotosProvider>
      </MessagesProvider>
    )
  }

  beforeEach(() => {
    jest.spyOn(api, 'editPhoto')
    jest.spyOn(api, 'getPhoto')
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should edit photo', async () => {
    const props = {
      onClose: jest.fn()
    }

    const { asFragment } = renderComponent(props)

    expect(asFragment()).toMatchSnapshot()
    expect(api.getPhoto).not.toHaveBeenCalled()

    fireEvent.change(screen.getByLabelText(/title/i), {
      target: { value: 'Photo title' }
    })

    fireEvent.change(screen.getByLabelText(/description/i), {
      target: { value: 'Photo description' }
    })

    await act(async () => {
      await fireEvent.change(document.querySelector('input[type="file"]'), {
        target: {
          files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
        }
      })
    })

    fireEvent.change(screen.getByLabelText(/position/i), { target: { value: 'left' } })

    fireEvent.click(screen.getByTestId('submit'))

    const formData = new FormData(document.querySelector('form'))

    expect(api.editPhoto).toHaveBeenCalledWith(198, formData)

    await waitFor(() => expect(props.onClose).toHaveBeenCalled())
  })

  it('should load photo', () => {
    const props = {
      id: 1
    }

    renderComponent(props)

    expect(api.getPhoto).toHaveBeenCalledWith(1)
  })

  it('should display detail', () => {
    const props = {
      detail: global.__PHOTO__,
      id: 1
    }

    const { asFragment } = renderComponent(props)

    expect(asFragment()).toMatchSnapshot()
  })
})
