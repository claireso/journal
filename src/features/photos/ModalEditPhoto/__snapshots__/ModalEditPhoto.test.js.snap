// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalEditPhoto /> should load photo 1`] = `
<DocumentFragment>
  <h1
    class="c-eskfFe"
  >
    Edit photo
  </h1>
  <form
    action=""
    enctype="multipart/form-data"
    method="POST"
  >
    <div
      class="c-dMoIJA"
    >
      <label
        class="c-iKAemi"
        for="title"
      >
        Title
      </label>
      <input
        class="c-hjpsDP"
        data-testid="title"
        id="title"
        name="title"
        type="text"
        value="Single photography"
      />
    </div>
    <div
      class="c-dMoIJA"
    >
      <label
        class="c-iKAemi"
        for="description"
      >
        Description
      </label>
      <input
        class="c-hjpsDP"
        data-testid="description"
        id="description"
        name="description"
        type="text"
        value="Janvier 2019"
      />
    </div>
    <div
      class="c-dMoIJA"
    >
      <label
        class="c-iKAemi"
        for="file"
      >
        Photo
      </label>
      <div
        class="c-iNomHb"
      >
        <div
          class="c-ktAHvd c-ktAHvd-itDbsJ-css"
        >
          <img
            src="/uploads/01d2tf2h38pwcd953ans2f64p7.jpg"
          />
        </div>
        <div
          class="c-cpxbDL"
        >
          <svg
            height="16"
            viewBox="0 0 26 28"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 23c0-0.547-0.453-1-1-1s-1 0.453-1 1 0.453 1 1 1 1-0.453 1-1zM24 23c0-0.547-0.453-1-1-1s-1 0.453-1 1 0.453 1 1 1 1-0.453 1-1zM26 19.5v5c0 0.828-0.672 1.5-1.5 1.5h-23c-0.828 0-1.5-0.672-1.5-1.5v-5c0-0.828 0.672-1.5 1.5-1.5h6.672c0.422 1.156 1.531 2 2.828 2h4c1.297 0 2.406-0.844 2.828-2h6.672c0.828 0 1.5 0.672 1.5 1.5zM20.922 9.375c-0.156 0.375-0.516 0.625-0.922 0.625h-4v7c0 0.547-0.453 1-1 1h-4c-0.547 0-1-0.453-1-1v-7h-4c-0.406 0-0.766-0.25-0.922-0.625-0.156-0.359-0.078-0.797 0.219-1.078l7-7c0.187-0.203 0.453-0.297 0.703-0.297s0.516 0.094 0.703 0.297l7 7c0.297 0.281 0.375 0.719 0.219 1.078z"
            />
          </svg>
          <span>
            Upload new photo
          </span>
          <small>
            (only jpg and png)
          </small>
          <input
            class="c-jdaWEC"
            name="file"
            type="file"
          />
        </div>
      </div>
    </div>
    <div
      class="c-dMoIJA c-TAzET"
    >
      <label
        class="c-iKAemi"
        for="position"
      >
        Position
      </label>
      <div
        class="c-gMvCio"
      >
        <select
          class="c-iGJQuL"
          id="position"
          name="position"
          required=""
        >
          <option
            value="left"
          >
            Left
          </option>
          <option
            selected=""
            value="center"
          >
            Center
          </option>
          <option
            value="right"
          >
            Right
          </option>
        </select>
        <svg
          height="14"
          viewBox="0 0 9 28"
          width="9"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.297 15c0 0.125-0.063 0.266-0.156 0.359l-7.281 7.281c-0.094 0.094-0.234 0.156-0.359 0.156s-0.266-0.063-0.359-0.156l-0.781-0.781c-0.094-0.094-0.156-0.219-0.156-0.359 0-0.125 0.063-0.266 0.156-0.359l6.141-6.141-6.141-6.141c-0.094-0.094-0.156-0.234-0.156-0.359s0.063-0.266 0.156-0.359l0.781-0.781c0.094-0.094 0.234-0.156 0.359-0.156s0.266 0.063 0.359 0.156l7.281 7.281c0.094 0.094 0.156 0.234 0.156 0.359z"
          />
        </svg>
      </div>
    </div>
    <input
      class="c-VRFby"
      data-testid="submit"
      type="submit"
      value="Save"
    />
  </form>
</DocumentFragment>
`;
